### ISSUE TITLE
Slow performance on large code blocks

### ISSUE THREAD

**Issue by modem7** [BookStackApp/BookStack#2518]
**Describe the bug**
On large code blocks (1000+ line), editing performance is quite slow, and takes a while to edit it. 

Can take around a minute to be able to start editing. 

CPU spikes to 18% when opening (not surprising) for about 1-2 seconds, and no RAM usage change (currently only using around 120MB). 

**Steps To Reproduce**
1. Create code block
2. Paste large amounts of data
3. Save
4. Re-open page, and edit code block

**Expected behavior**
Potentially faster/better optimised code block editing (if possible?)

**Screenshots**
https://omegawiki.modem7.com/books/docker/page/current-docker-compose

**Your Configuration (please complete the following information):**
 - Exact BookStack Version (Found in settings): v0.31.4
 - PHP Version: Whatever is in the LinuxIO image
 - Hosting Method (Nginx/Apache/Docker): Docker
 - Browser: Vivaldi (Chrome based)

**Additional context**
This may be an unusual edge case, so absolutely understandable if it's not high up on the priority list to investigate.

**Comment by ssddanbrown** [comment]
Hi @modem7,

Could you confirm what editor you're using? The Default WYSIWYG editor or the markdown editor?

Additionally, Could you confirm the exact editing step that you find troubling slow? Opening the code block for edit? Making the change? Saving the change?

**Comment by modem7** [comment]
Heya, 

WYSISYG editor in this case. 

And to better explain: https://www.dropbox.com/s/yp6sn2jwsxacyk0/Current%20Deployment%20_%20OmegaWiki%20-%20Vivaldi%202021-01-30%2015-41-29.mp4 I made a screenshare video

**Comment by ssddanbrown** [comment]
Thanks for confirming @modem7. Looks like the code editor popup window performs okay, but the main editor is super slow.

---

### Initial Findings

Chrome performance chart, Page editor load, 4x CPU slowdown:

![image](https://user-images.githubusercontent.com/8343178/106383415-b87b7700-63bd-11eb-917a-477e38ab9743.png)

Green parts are CodeMirror which represents most of the code execution time, Can see two major events which consume a lot fo time.

Sequence of events appears as follows:

- TinyMCE loaded.
- TinyMCE content set.
   -  During TinyMCE initial set content, CodeMirror loaded for each code section.
      - CodeMirror renders out code sections, **_Potential duplicate work three very similar long-running patchDisplay calls_**. 
      - CodeMirror `updateHeightsInViewport` call runs causing massive style/layout shift. **_Very Heavy_** 
- TinyMCE adds page styles, **_Causes massive style/layout calculation_**  
- TinyMCE finishes up _(Note, Mulitple editor change events fired in this process causing slowdown in calls to getContent for saving.)_
- `cm.refresh()` call from BookStack WYSIWYG code function runs after timeout, causing re-run of all rendering and layout calculations. **_Very Heavy Again_** 

There's duplication in CodeMirror block rendering here. Both before and after editor styles loaded. Should probably render codemirror on set content, as we're doing, but defer visibility and display of CodeMirror content until after styles loaded. Potentially test with CodeMirror in its own forced rendering layer, if not already, to see if that offers a performance boost.

**Comment by ssddanbrown** [comment]
### Chart after improvements

Chrome performance chart, Page editor load, 4x CPU slowdown:

![image](https://user-images.githubusercontent.com/8343178/106390655-56346d80-63e1-11eb-8ab2-483cbf336d79.png)

**Comment by ssddanbrown** [comment]
Hi @modem7,

I've now had a dive into this and explored the situation, There's a couple of things at play:

1. Inefficient loading of code blocks into the editor causing duplicate browser work in processing styles and layout.
2. Code block performance is poor with long-content blocks due to needing to re-calculate height to show all content on the page.

For 1, I have patched the code in bc1e84325c7edfb52b07c7e6bde881b09280c3ba, as detailed above, to reduce the inefficient duplicated work. This reduced loading down to 60% of the original time from my testing, which may vary depending on device and content. 

For 2, there's not much I can do within the code based without altering how people's page content displays although you can apply a workaround on your own instance if desired. Limiting the height of code blocks brings massive performance benefits in these situations. You can do this by adding the following to your "Custom HTML Head Content" setting:

```html
<style>
.CodeMirror, .CodeMirror-scroll {
  max-height: 900px;
}
</style>
```

That should improve both page viewing and editing performance dramatically. 

The fixes for 1. above will be part of the next feature release. Since I've addressed the main BookStack inefficiency ready for next release, and provided a workaround for the remainder of the performance issues, I'll close this off but feel free to comment here still if required.

**Comment by modem7** [comment]
Thank you very much! I'll apply the workaround and wait for the rollout! 

Appreciate the quick turnaround on this, thanks again!
